### PR TITLE
FIX: Ensure video tags with autoplay are muted

### PR DESCRIPTION
**Pull request by xfalcox**

### CODE REVIEW

File: app/assets/javascripts/pretty-text/addon/sanitizer.js [hunk 1]
```diff
@@ -143,6 +143,11 @@ export function sanitize(text, allowLister) {
         }
       }
     },
+    onTag(tag, html, options) {
+      if (tag === 'video' && html.match(/autoplay/) && !html.match(/muted/)) {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/pretty-text/addon/sanitizer.js:147]: Hmmm... what happens if the video src includes the words 'autoplay' or 'muted' 🤔 Ideally we would do some proper HTML parsing here, but I'm not sure we have the functions available to do it...

File: app/assets/javascripts/pretty-text/addon/sanitizer.js [hunk 2]
```diff
@@ -143,6 +143,15 @@ export function sanitize(text, allowLister) {
         }
       }
     },
+    onTag(tag, html) {
+      if (
+        tag === "video" &&
+        html.match(/ autoplay/i) &&
+        !html.match(/ muted/i)
+      ) {
+        return "[removed]";
```
Comment (riking) [inline app/assets/javascripts/pretty-text/addon/sanitizer.js:152]: Can you automatically add the muted attribute instead?

Comment (xfalcox) [reply app/assets/javascripts/pretty-text/addon/sanitizer.js:152]: That is possible and I guess it's better?

**Review by davidtaylorhq — APPROVED** [review]

**Review by davidtaylorhq — CHANGES_REQUESTED** [review]
I think this would sneak through

```
<video autoplay=" muted">
```
I don't think solving this securely with regex is going to be easy 😢

**Comment by xfalcox** [comment]
> I think this would sneak through
> 
> ```
> <video autoplay=" muted">
> ```
> 
> I don't think solving this securely with regex is going to be easy - we need to parse the html attributes according to the spec. Maybe we can pull something out of the `xss.js` library to help with that parsing?

Yeah, the XSS library doesn't really helps with having the current tag and each parsed attribute at all...

**Comment by davidtaylorhq** [comment]
@xfalcox it's slightly misusing the library, but I think this kind of thing could work:

```javascript
onTag(tag, html, options) {
  if (tag === "video" && !options.isClosing) {
    // One pass to check for a muted attribute
    let attrs = [];
    xss(html, {
      onTagAttr: (_, name) => attrs.push(name),
    });

    const addMute = attrs.includes("autoplay") && !attrs.includes("muted");
    if (!addMute) {
      return;
    }

    // Sanitize as normal, but add muted attribute
    return xss(html, {
      whiteList: allowList.tagList,
      onTagAttr: (_, name) => {
        if (name === "autoplay") {
          return "autoplay muted";
        }
      },
    });
  }
},
```

**Comment by xfalcox** [comment]
Force pushed with your proposed approach plus tests @davidtaylorhq. Great idea on the way you were able to re-use the standard `onIgnoreTagAttr` callback with two attributes!

File: app/assets/javascripts/pretty-text/addon/allow-lister.js [hunk 3]
```diff
@@ -225,8 +225,7 @@ export const DEFAULT_LIST = [
   "track[srclang]",
   "ul",
   "video",
-  "video[autoplay]",
-  "video[controls]",
+  // video[autoplay] handled by sanitizer.js  "video[controls]",
```
Comment (davidtaylorhq) [inline app/assets/javascripts/pretty-text/addon/allow-lister.js:228]: I think we lost a linebreak here

**Review by davidtaylorhq — APPROVED** [review]

**Pull request merged**